### PR TITLE
Add splash image rendering with admin and backend functions

### PR DESCRIPTION
**Pull request by jfietkau**
Okay, here's the big one.

This PR adds splash image rendering for starter packs using Pillow. Splash images are generated by fetching up to 24 avatars of accounts in the starter pack, arranging them into a nice hex grid, adding the title of the pack and saving everything into a local media directory.

Besides the splash image rendering itself, this PR contains an admin function that allows you to (re-)render splash images for specific starter packs in the admin view, as well as a backend command `update_splash_images` that checks for starter packs needing splash image updates and handles them in the background. It also contains data model changes that mark starter packs where a splash image rerender *would* be required (e.g. after adding or removing accounts in a way that would influence the splash image content), but do not automatically act on these markings.

This PR does **not** yet use the rendered starter packs in the UI or triggers any automatic render processes outside of the two administrative functions above (although the generated splash images are technically web-accessible if you know the path). This version is intended to test the feature's resource usage and resilience under real circumstances.

This PR also does not contain any unit tests. Graphics rendering is difficult to unit test (do we want to compare pixel color values?), and outside of graphics, network, and storage output, you could at best test details around the margins (e.g. "the number of avatars on a splash image is <= the number of accounts in the starter pack) for which the implementation is mostly trivial. Does this need to change, should we have unit tests for splash images and if so, what should they look like?

I'm going to leave a few architectural concerns as review comments myself. Open to any required changes as always.

### CODE REVIEW

File: starter_packs/views.py [hunk 1]
```diff
@@ -671,3 +690,303 @@ def follow_bg(user_id: int, starter_pack_slug: str):
...
+    accounts = get_splash_image_accounts(starter_pack)
+    random.shuffle(accounts)
+
+    inner_width = render_resolution[0] - 2 * margin_side
+    inner_height = render_resolution[1] - margin_top - margin_bottom
+    inner_aspect_ratio = inner_width / inner_height
+
+    num_lines = 1 + math.floor(math.sqrt(max(len(accounts) - 1, 1) / inner_aspect_ratio))
+    per_line = math.floor(len(accounts) / num_lines)
+    num_lines = math.floor(len(accounts) / per_line)
```
Comment (anze3db) [inline starter_packs/views.py:873]: If a starter pack doesn't have any accounts this line errors out with:

```
  File "/Users/anzepecar/code/fedidevs/starter_packs/management/commands/update_splash_images.py", line 82, in render_splash_images_for
    render_splash_image(starter_pack, host_attribution)
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/anzepecar/code/fedidevs/starter_packs/views.py", line 873, in render_splash_image
    num_lines = math.floor(len(accounts) / per_line)
                           ~~~~~~~~~~~~~~^~~~~~~~~~
ZeroDivisionError: division by zero
```

File: starter_packs/views.py [hunk 2]
```diff
@@ -671,3 +690,303 @@ def follow_bg(user_id: int, starter_pack_slug: str):
...
+            # already exists and is _not_ a directory.
+            logger.exception("Unable to create %s directory: %s", directory, type(e).__name__)
+            return
+        try:
+            write_test_path = directory / "__write_test.lock"
+            write_test_path.touch()
+            write_test_path.unlink()
+        except OSError:
+            logger.exception("Unable to write to directory %s", directory)
+            return
```
Comment (anze3db) [inline starter_packs/views.py:846]: I think there is no need for trying to create/test dirs every time we run `render_splash_image`. I think Django will take care of dir creation for us and if it doesn't I can create them manually once and it should work from then on.

```suggestion
```

File: starter_packs/views.py [hunk 3]
```diff
@@ -671,3 +690,303 @@ def follow_bg(user_id: int, starter_pack_slug: str):
...
+    )
+
+    empty_avatar = Image.new("RGBA", (crop_mask.width, crop_mask.height), (0, 0, 0, 0))
+    avatars = []
+    for account in accounts:
+        avatar = fetch_avatar(account.avatar, crop_mask)
+        if avatar is not None:
+            avatars.append(avatar)
+        else:
+            logger.warning("Failed to fetch avatar of account %s for splash image", account.get_username_at_instance())
```
[the remark this answers is not in the document]
Comment (anze3db) [reply starter_packs/views.py:913]: I'm getting a few unreachable avatars when testing this locally :(

Image

Since the number of accounts in a starter pack is limited we can fetch all of them in `get_splash_image_accounts` and then start fetching avatars until we either run out of accounts or `SPLASH_IMAGE_NUMBER_OF_AVATARS` is reached. Since we have limits in place for downloading I'm not too concerned about memory issues from having multiple avatars in memory but if this does at some point become an issue we could also dump them to disk in this step and re-read when it's time to insert them back.

File: starter_packs/views.py [hunk 4]
```diff
@@ -671,3 +690,303 @@ def follow_bg(user_id: int, starter_pack_slug: str):
...
+    a starter pack has been modified, or if the content is unchanged under
+    the current render settings.
+    """
+    accounts = get_splash_image_accounts(starter_pack)
+    accounts = [account.get_username_at_instance() for account in accounts]
+    content = starter_pack.title + ",".join(sorted(accounts))
+    return hashlib.sha512(content.encode("utf-8")).hexdigest()[:32]
+
+
+def get_splash_background(width, height, attribution, font, cache_path):
```
[the remark this answers is not in the document]
Comment (anze3db) [reply starter_packs/views.py:728]: We could also commit the generic splash background image background to code. I don't think it really needs to be generated but this also works!

File: starter_packs/views.py [hunk 5]
```diff
@@ -671,3 +690,303 @@ def follow_bg(user_id: int, starter_pack_slug: str):
...
+def get_splash_image_accounts(starter_pack):
+    accounts = (
+        Account.objects.filter(
+            starterpackaccount__starter_pack=starter_pack,
+            instance_model__isnull=False,
+            instance_model__deleted_at__isnull=True,
+            discoverable=True,
+        )
+        .select_related("accountlookup", "instance_model")
+        .order_by("starterpackaccount__created_at")
```
[the remark this answers is not in the document]
Comment (anze3db) [reply starter_packs/views.py:709]: Perfect 🎉

File: starter_packs/views.py [hunk 6]
```diff
@@ -671,3 +690,303 @@ def follow_bg(user_id: int, starter_pack_slug: str):
 
             logger.error("Unknown error when following %s", account.username_at_instance)
             continue
+
+
+# The maximum number of avatars that will appear on a starter pack splash
```
[the remark this answers is not in the document]
Comment (anze3db) [reply starter_packs/views.py:695]: Personally I would add the splash image generation code into something like: starter_packs/splash_images.py

But this file isn't too big to scroll through yet so I also wouldn't mind keeping everything here for now.

File: starter_packs/management/commands/update_splash_images.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        )
+
+    def main(
+        self,
+        stale_after: int = 0,
+        seconds_between: int = 0,
+        host_attribution: str = "",
+        run_forever: bool = False,
+    ):
+        # This command works synchronously and in a single thread in order to avoid
```
[the remark this answers is not in the document]
Comment (anze3db) [reply starter_packs/management/commands/update_splash_images.py:46]: I think this is perfectly fine for now. If it gets too slow I would probably try to parallelize fetching the avatars themselves and not this main function as a whole. We can parallelize avatar fetching per instance so that we never hit a single server with too many requests at once although CDN servers that usually host avatars should generally not have issues with multiple requests at once.

**Review by anze3db — COMMENTED** [review]
I left a few notes. The division by zero error below is the main thing blocking me from merging this.

Regarding testing:

I wouldn't assert visual elements of the generated image but I would have a test case or two that goes through the code and makes sure the image gets generated without errors - just so that we can test edge cases like trying to generate a splash image without any accounts.

Regarding the style:

When there are one or two accounts the circles feel a bit too big to me:

Image

I think it would look much better if the avatars never took over more than 1/3 of the whole image and the title and hosted by text could also use a bit more padding.

Again I think Blusky nailed their design and the amount of padding that they use around text/images:

Image

File: starter_packs/views.py (hunk 1 above)
Comment (jfietkau) [reply starter_packs/views.py:873]: Fair, I didn't test that case.

Thinking about it, what *should* we do if a starter pack has no accounts in it? We could render an image with just the title and gradient background, but it would look very empty. Or we could have `render_splash_image` fail fast so that the starter pack will have no splash image and the UI and API can fall back to the generic one.

File: starter_packs/views.py (hunk 2 above)
Comment (jfietkau) [reply starter_packs/views.py:846]: Django may create `media` if it is smart enough to deduce it from the settings, but it will definitely not create `media/splash` since that's only used programmatically here.

I did the write test at the start so we could fail fast if something is wrong with the directories, but if we remove the check here, what ends up happening is that we need to wrap the file writes at the ends of `render_splash_image` and `get_splash_background` in `try ... except` instead. I have no problem with that, the only downside is if there is a problem with the file system after all, we may end up downloading the avatars and generating the splash image before we realize we can't save the result.

File: starter_packs/views.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jfietkau) [reply starter_packs/views.py:913]: Changed the background for avatars with transparency to a lighter gray that is semi-transparent. Should look better for almost all affected avatars I think.

**Comment by jfietkau** [comment]
I can't help but feel that the style feedback would have been easier to incorporate at the mock-up stage. 😅 I've capped the avatar size to 1/3 of the image like you suggested, and changed the title position to be dynamic based on the total height of the avatar area. I hope it looks good for larger numbers of avatars than in my local test starter packs.

The host attribution can't easily depend on the avatar area though, since it's part of the background. Tbh it's kinda hidden away in the corner on purpose. When you say more padding, how much do you mean? Would you center it like in the Bluesky splash image? As a matter of pride, I don't want to copy them 1:1.

**Comment by jfietkau** [comment]
For emoji support, it is indeed non-trivial, but we have a few options. Pillow doesn't support color font rendering at all yet, so the solutions confront that issue in different ways:

1. We could add monochrome TrueType emoji ([these bad boys](https://fonts.google.com/noto/specimen/Noto+Emoji)) by picking an open emoji font and mixing that in. The simplest way would be to use fontforge to merge Inter and whatever emoji font we pick into a new combined TTF file. I haven't tried it, but I think Pillow should just be able to use it to render outlined emoji. The aesthetic isn't what most people expect in 2025, but it's arguably better than removing them entirely, which might change the meaning of some titles.
2. We could use [pilmoji](https://pypi.org/project/pilmoji/). I tried it out and it has a number of downsides: it introduces a network dependency on [emojicdn.elk.sh](https://emojicdn.elk.sh) into text rendering, the image resolution of the resulting emoji looks grimy in our titles, plus there's an incompatibility between all released versions of pilmoji and the underlying emoji library it uses, which would mean we'd have to pin the dependency to an old version so pilmoji can run. A fix for that was merged into pilmoji's `main` a year ago, but the last versioned release is from 2023, so I'm not sure what exactly is going on over there.
3. I could try to roll our own bitmap emoji rendering by bundling a full set of Twemoji (or something) and doing what pilmoji does. It looks non-trivial but not insurmountable.
4. Or we could strip them out as you suggest, that can probably be done by looking at the Unicode character ranges.

Leaning towards 1.

**Comment by jfietkau** [comment]
I have added monochrome emoji suport to starter pack titles in splash images. I ended up not combining the font files directly and instead splitting title segments based on character types, this adds a bit of code complexity but it keeps font copyrights and licensing clearer.

It raised the question for me what our requirements are for internationalization. Inter doesn't have a ton of non-latin alphabets it seems, so most other writing systems (Japanese, Chinese, Arabic...) currently won't display because Pillow can't fall back to supported fonts the way browsers can. If we want to solve this, maybe I can reuse the branching logic to swap from Inter to Noto for characters that Inter doesn't support.

**Comment by jfietkau** [comment]
I've added support for arbitrary fonts in splash image starter pack titles, font switching depending on glyph availability, and a handful of Noto Sans files for different alphabets. This isn't exhaustive, [there's a lot of them](https://fonts.google.com/noto/fonts?query=noto%20sans). But it probably makes some sense to check if we're still missing anything (unprintable characters now log a warning) before preemptively bloating the repo. Seems to work for my testing.

test

**Comment by jfietkau** [comment]
Thank you for merging! I hope it goes well, looking forward to the results. I'm sure there's quite the backlog of starter packs to go through. You can let me know here or in a new issue if there's anything that needs a code fix.

**Comment by anze3db** [comment]
It went well and there were no issues. All starter packs in prod have been generated 🎉
